Typescript - Final PR

NoraCodes opened the pull request.

NoraCodes committed:
Add TypeScript build pipeline

Modify files necessary to build app with TypeScript.

The app is working in dev mode as of this commit.

NoraCodes committed:
Move front-end files to TypeScript extension

and make necessary changes, mostly in imports, to make the app build.

NoraCodes committed:
Fix type errors with lax TypeScript configuration.

Note: This required changing the export mechanism for TimeInterval, which required changing a file on the server. Functionality is not affected on the frontend; all backend tests pass.

NoraCodes committed:
Enable strict checks for TypeScript

NoraCodes committed:
Merge changes from groupUI into typing branch.

NoraCodes committed:
Move files to .ts/tsx and enforce

NoraCodes committed:
Absolute minimum changes to get app working post GroupUI merge.

Lots of type errors, still.

NoraCodes committed:
First run at fixing groupUI type errors.

Also fix a bug where you couldn't submit groups that contain meters?? See DatasourceBoxComponent, it was checking for strings that were never passed.

NoraCodes committed:
Fully typed Redux state.

NoraCodes committed:
Tons and tons of TypeScript!

This commit also adds a lot of comments of the form TODO TYPESCRIPT: <something I couldn't figure out or which will take time>.

NoraCodes committed:
Fix all fixable client-side TS issues.

Server-side lint failures still exist.

8 client-side type errors exist. They are marked with TODO TYPESCRIPT.

NoraCodes committed:
Merge 'master' into typescript

NoraCodes committed:
Fix type errors with FormControl events.

NoraCodes committed:
Fix all but one TSLint error.

NoraCodes committed:
Fix final linter error.

NoraCodes committed:
Add typeCheck NPM script.

NoraCodes committed:
Fix bug where we were submitting a whole group instead of its ID.

This highlights the need to typescriptify the backend as well, as soon as feasible. Then, we can share types between them and catch errors like this in the typechecking stage.

NoraCodes committed:
Address PR concerns:

* re-format babelrc
* remove misleading reference to eslint
* re-organize and re-format installation and usage instructions
* re-organize TimeInterval type definition
* fix indentation and style in several places
* Comment TSLint config
* Fix spelling errors in bar readings actions
* Don't reduntantly define ActionType strings for group actions
* Remove some unused imports
* Remove outdated JSDoc
* Properly type LineReading actions
* Resolve type issue where barDuration was sometimes a number and sometimes a Duration
* Resolve type issue where compareTimeInterval was sometimes a number and sometimes a TimeInterval. THIS BREAKS COMPARE but only because it was already broken; rewrite is in the works.

NoraCodes committed:
Big Typescript reorganization re: John's comments.

johndiiorio committed:
Merge development into typescript-preferences, begin converting preferences to typescript

johndiiorio committed:
Finish typing preferences in Redux, move graph types to redux type file

johndiiorio committed:
Type Admin and Header component/container

johndiiorio committed:
Fix typings in GroupSideBarComponent, AdminComponent. Move all fetches on componentWillMount to InitializationComponent

johndiiorio committed:
Merge branch 'typescript' into typescript-preferences

# Conflicts:
#	src/client/app/actions/graph.ts
#	src/client/app/actions/groups.ts
#	src/client/app/components/LoginComponent.tsx
#	src/client/app/components/RouteComponent.tsx
#	src/client/app/containers/ChartDataSelectContainer.ts
#	src/client/app/containers/LoginContainer.ts
#	src/client/app/containers/RouteContainer.ts
#	src/client/app/containers/UIOptionsContainer.ts
#	src/client/app/containers/groups/GroupSidebarContainer.tsx
#	src/client/app/reducers/barReadings.ts
#	src/client/app/reducers/graph.ts
#	src/client/app/reducers/index.ts
#	src/client/app/types/redux/actions.ts

johndiiorio committed:
Lots of minor type fixing, standardize MPL header format

johndiiorio committed:
Fix improper paths to types

johndiiorio committed:
PR fixes. Fix regression where group edits are not viewable without refresh

NoraCodes committed:
Fix remaining resolvable type errors.

johndiiorio committed:
Merge pull request #258 from LeoTindall/typescript-preferences

Fix remaining resolvable type errors for Typescript Preferences branch

NoraCodes committed:
Merge pull request #5 from OpenEnergyDashboard/typescript-preferences

Typescript preferences

NoraCodes committed:
Add Compare actions

NoraCodes committed:
Update TypeScript to development.

The remaining type errors are issues with type information.
One should be resolved soon.

NoraCodes committed:
Merge branch 'development' of github.com:OpenEnergyDashboard/OED into typescript

NoraCodes committed:
Fix linting errors.

johndiiorio committed:
Merge pull request #297 from OpenEnergyDashboard/development

Update master to 0.3.0

simonbtomlinson committed:
Merge branch 'development' into typescript

NoraCodes committed:
Squash a lot of type errors.

NoraCodes committed:
Merge development into TypeScript post 0.3.0

NoraCodes committed:
Fix visibility of shouldComponentUpdate()

johndiiorio commented on package.json:
```diff
@@ -7,8 +7,11 @@
     "start": "node ./src/bin/www",
     "dev": "webpack -d --watch --color --progress",
     "build": "cross-env NODE_ENV=production webpack -p",
+    "buildDev": "webpack -p",
```
We should be using the `build` script instead of what `buildDev` does, so we should delete this line.

johndiiorio commented on .travis.yml:
```diff
@@ -36,7 +36,9 @@ before_script:
   - npm run checkHeader
-    # Check for eslint failures
+    # Validate that all areas which are intended to be typed, are typed
+  - npm run checkTypescript
```
Should we also run `npm run typeCheck`?

NoraCodes replied:
We couldn't while there were still unresolvable type errors; thanks to your PR we now can.

johndiiorio replied:
If you agree, I'd like to rename `checkTypescript` to something more descriptive. It sounds too similar to `typeCheck`.

johndiiorio commented on src/client/app/actions/groups.ts:
```diff
@@ -316,14 +272,17 @@ function submitGroupEdits(group) {
-			const rawGroup = getState().groups.groupInEditing;
+			const rawGroup = getState().groups.groupInEditing as t.GroupDefinition;
+			if (rawGroup === undefined) {
+				throw new Error('Unacceptable condition: state.groups.groupInEditing has no data.');
```
I would like to remove all the `throw new Error` statements on the client, or at the very least display an error to the user. I don't think they serve much purpose except for debugging purposes, which should not be in production.

In this case, I would just do:
```
if (rawGroups) {
   ...
}
```

simonbtomlinson replied:
I highly disagree with you. We shouldn't support our application when it's in weird bad states. Failing fast is always better as it's easier to debug (and at some point we will need to debug something in production), and as the user is just going to have the application subtly break on them later when the application accesses `undefined.startTimestamp` or something. Catching and doing nothing doesn't fix the problem, it just hides it, and it does a bad job of it.

jameeters replied:
And for what it's worth, unless this has been fundamentally changed for typescript, the application will not recover from an empty group in editing marked dirty, and that's the one case that will lead to this error condition. The app will stop working properly whether or not we throw an error.

johndiiorio commented on src/client/app/actions/groups.ts:
```diff
@@ -343,11 +302,15 @@ export function submitGroupInEditingIfNeeded() {
 	return (dispatch, getState) => {
 		dispatch(markGroupInEditingDirty());
+		const groupInEditing = getState().groups.groupInEditing as t.GroupDefinition;
+		if (groupInEditing === undefined) {
```
Again, another instance of the `throw new Error`

jameeters replied:
Again, this state is not recoverable. The application will not behave correctly once in this state whether or not an error is thrown. An error is the best way to guarantee that things fail as cleanly as possible.

johndiiorio commented on src/client/app/components/DashboardComponent.tsx:
```diff
@@ -0,0 +1,47 @@
+
+/* tslint:disable no-string-literal */
+// TODO TYPESCRIPT: There is an open PR for this: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/21842
+defaults.global.plugins = {datalabels: {display: false}};
```
This PR has been merged. Is this still an issue, and if so, we should comment it differently.

NoraCodes replied:
I need to check that, but I think it is fixed and I'll remove the comment.

johndiiorio commented on src/client/app/components/InitializationComponent.tsx:
```diff
@@ -0,0 +1,48 @@
+
+	public componentWillMount() {
+		this.props.fetchMeterDetailsIfNeeded();
+		this.props.fetchGroupDetailsIfNeeded();
```
This line is not part of development. Why do we fetch group details here?

NoraCodes replied:
I experienced an issue with not having groups loaded and adding this line fixed that.

johndiiorio commented on src/client/app/components/InitializationComponent.tsx:
```diff
@@ -0,0 +1,48 @@
+		}
+	}
+
+	public shouldComponentUpdate() {
```
This function has been moved into RouteComponent (where it should go). Thus, it is safe to delete this function.

NoraCodes replied:
Thanks! As with all things, feel free to make and commit that change; you can push to my fork, afaik

johndiiorio commented on src/client/app/components/MultiSelectComponent.tsx:
```diff
@@ -2,39 +2,45 @@
 		// selectedOptions holds a list of the options that have been selected
 		this.state = {
-			selectedOptions: this.props.selectedOptions ? this.props.selectedOptions : null
+			selectedOptions: this.props.selectedOptions ? this.props.selectedOptions : undefined
```
This should definitely be null.

NoraCodes replied:
Could you elaborate?

johndiiorio replied:
There was a bug in MultiSelectComponent that @sandeepacharya464 and I had when `selectedOptions: undefined`, which was fixed by setting `selectedOptions: null`. I believe it had to do with the underlying Select component's PropTypes.

NoraCodes replied:
Thank you.

NoraCodes replied:
Actually, after looking into this, that's not consistent with the type signature of `Select` (from `react-select`), and it doesn't cause a problem, so I don't see the need to change it.

johndiiorio commented on src/client/app/components/RouteComponent.tsx:
```diff
@@ -73,14 +78,15 @@ export default class RouteComponent extends React.Component {
 							break;
 						case 'barDuration':
-							options.barDuration = moment.duration(parseInt(info), 'days');
+							const maybeDuration = moment.duration(parseInt(info));
```
I think there was a mistake here. This should definitely be assigned to `options.barDuration` as maybeDuration is never used. Should the `days` parameters still be removed?

NoraCodes replied:
Fixed; good catch, thank you.

johndiiorio commented on src/client/app/components/RouteComponent.tsx:
```diff
@@ -73,14 +78,15 @@ export default class RouteComponent extends React.Component {
 						case 'barStacking':
 							if (this.props.barStacking.toString() !== info) {
-								options.changeBarStacking = true;
+								options.toggleBarStacking = true;
```
GitHub doesn't let me select the line 4 down from here, but it is another case of the client-side erroring. Ideally this could be fixed, but not critical.

johndiiorio commented on src/client/app/components/RouteComponent.tsx:
```diff
@@ -73,14 +78,15 @@ export default class RouteComponent extends React.Component {
 						case 'chartType':
-							options.chartType = info;
+							options.chartType = info as ChartTypes;
+							options.chartType = info as ChartTypes;
```
Duplicated lines here

johndiiorio commented on src/client/app/components/UIOptionsComponent.tsx:
```diff
@@ -214,4 +178,71 @@ export default class UIOptionsComponent extends React.Component {
+	/**
+	 * Called when the user releases the slider, dispatch action on temporary state variable
+	 */
+	// TODO TYPESCRIPT: react-rangeslider doesn't have types. We should investigate what type this actually is
```
Recommend switching out to use `rc-slider`, which is strictly better and should have good typings. The change should also be very fast.

NoraCodes replied:
I've been working on this and unfortunately rc-slider is very broken; I can't get it to slide. If you could do this I would appreciate it a lot.

johndiiorio replied:
I will look into this.

johndiiorio commented on src/client/app/components/UIOptionsComponent.tsx:
```diff
@@ -214,4 +178,71 @@ export default class UIOptionsComponent extends React.Component {
+		this.props.changeBarStacking();
+	}
+
+	// TODO TYPESCRIPT this is an issue with typings for React.FormEvent<> and ChangeEvent<>
```
This "might" (not sure) be fixed in React 16+. Recommend investigating and possibly upgrading to React 16.

NoraCodes replied:
Agreed, I would like to do that.

johndiiorio commented on src/common/TimeInterval.js:
```diff
@@ -16,6 +16,9 @@ class TimeInterval {
 		if (this.isBounded) {
 			// Using '_' as a separator character since it doesn't appear in ISO dates
+			if (this.startTimestamp === undefined || this.endTimestamp === undefined) {
+				throw Error('startTimestamp or endTimestamp was undefined in a bounded TimeInterval');
```
Recommend using console.error here.

jameeters replied:
As long as that's in addition to throwing the error.

NoraCodes replied:
Can I ask why? What does it get us?

jameeters replied:
Are you asking me why I want the error or John why he wants the console log?

NoraCodes replied:
Why John wants the console error. Thrown errors get logged if unhandled.

johndiiorio commented on src/common/TimeInterval.js:
```diff
@@ -16,6 +16,9 @@ class TimeInterval {
 	toString() {
 		if (this.isBounded) {
 			// Using '_' as a separator character since it doesn't appear in ISO dates
+			if (this.startTimestamp === undefined || this.endTimestamp === undefined) {
```
Does it even make sense to throw an error here? Shouldn't we be able to handle half-open time intervals.

jameeters replied:
It absolutely makes sense to handle half-bounded intervals, but they should be considered unbounded. So a bounded time interval must have two endpoints, and if either of them is missing we have an error condition.
I don't think we need half-bounded intervals now (if we did, we'd have them) so this seems correct to me.

johndiiorio replied:
Makes sense, I agree.

johndiiorio commented on src/common/TimeInterval.js:
```diff
@@ -69,4 +72,4 @@ class TimeInterval {
 }
 
-module.exports = TimeInterval;
+module.exports = { TimeInterval };
```
We should be able to default export the TimeInterval class, e.g. `module.exports = TimeInterval`.

NoraCodes replied:
You already asked me about this in person, and already I told you that it doesn't work properly. I can't figure out a way to get it to work like that, but if you can I'd love to get this back to a default export.

johndiiorio reviewed (changes requested):
Initial pass. Nothing too critical here.

johndiiorio commented on USAGE.md:
````diff
@@ -57,7 +57,11 @@ OED_LOG_FILE=?                 // Path to the log file, defaults to ./log.txt
 1. Run ```systemctl enable oed.service``` to make the service start on server boot.
 1. Bring the app online with ```systemctl start oed.service```. Stop the app with ```systemctl stop oed.service```.
 
-### Administration ###
````
You changed this file to USAGE.md but forgot to update the link in the README

NoraCodes replied:
Fixed

johndiiorio committed:
Replace react-rangeslider with rc-slider

NoraCodes committed:
Merge pull request #6 from OpenEnergyDashboard/typescript-rc-slider

Replace react-rangeslider with rc-slider

NoraCodes committed:
Resolve PR issues.

NoraCodes committed:
Fix issue with incorrectly retrieving compare vals

NoraCodes committed:
Revert to week-long comparisons by default

johndiiorio commented:
Sorry about the merge conflicts. However, with the recent merge (which should help migrations), we should add documentation about the new `editUser` npm script. It doesn't need to be in this PR.

simonbtomlinson committed:
Fix selected group invalid after creating group

The old selected group's name was still displayed after creating a new
group. This clears the list of selected groups when switching
screens, which deselectes the group.

NoraCodes committed:
Merge branch 'master' of github.com:OpenEnergyDashboard/OED into typescript

NoraCodes committed:
Update to development

NoraCodes committed:
Fix chart hotlinking again

NoraCodes commented:
We're mergable!

Tested working:                                 
- Mobile layout                             
- Line chart & bar chart                    
- Creating Groups                           
- Editing groups                            
- Editing preferences                       
- Compare (all intervals)                   
- Import meter data                         
- CreateUser, UpdateMamacMeter, etc scripts 
- Chart Hotlinking

NoraCodes committed:
Fix lint errors in user utils

NoraCodes committed:
Prevent tree-shaking from removing plugins.

johndiiorio reviewed (approved):
Great job!
